claimRequest() and other powerbox-related updates

dwrensha opened the pull request:
This pulls together the work from #2016 and #2019.

The ip-networking integration tests are working after applying https://github.com/zarvox/sandstorm-test-python/pull/4.

I have not yet updated the other powerbox integration tests, which are based on https://github.com/jparyani/sandstorm-test-app. So this will not pass Jenkins.

I've split out a new file powerbox.capnp. Doing so isn't strictly necessary for this change, but if we want to do that eventually, now seems like an opportune time.

I've not yet addressed all of the comments in #2019, but I wanted to get this pull request up for discussion.

dwrensha commented:
@kentonv [this](https://github.com/sandstorm-io/sandstorm/blob/b7bad661698b561439491622df37c9e1d6a18386/shell/server/core.js#L498-L507) is the code that I was talking about earlier today. It sets the `owner.user.title` field based on the `owner.grain.saveLabel` field, which feels wrong to me. In my understanding, `saveLabel` is used by the app to explain to the user why it saved the capability. I can imagine it being weird if such an explanation ended up in a user's title for a grain.

kentonv commented:
@dwrensha I agree, that does look wrong. It seems like the code meant to use `tokenInfo.owner.user.title`.

dwrensha commented:
Hm... it also occurs to me that `ApiTokenOwner.grain.introducerIdentity` is obsolete after this change. It's role gets fulfilled instead by `ApiTokenOwner.clientPowerboxRequest.introducerIdentity`.

zarvox commented:
> Hm... it also occurs to me that `ApiTokenOwner.grain.introducerIdentity` is obsolete after this change. It's role gets fulfilled instead by `ApiTokenOwner.clientPowerboxRequest.introducerIdentity`.

I'm not so sure.

Would making that change require that other users who want to review/audit the capabilities owned by a grain also be able to see `ApiTokens` with owner `clientPowerboxRequest`?  I'm thinking about what information you'd need to build a compelling "audit connections" UI.

Does this still make sense if the powerbox request is initiated server-side, or do we intend to drop that flow entirely?  Seems we'd need a place for the introducer in that workflow too, but you wouldn't necessarily want to chain to a `clientPowerboxRequest` there.

dwrensha commented:
> Would making that change require that other users who want to review/audit the capabilities owned by a grain also be able to see ApiTokens with owner clientPowerboxRequest? I'm thinking about what information you'd need to build a compelling "audit connections" UI.

No, `clientPowerboxRequest` tokens are short-lived and would not need to show up in an audit UI.

The `introducerIdentity` field is only needed to determine which user to apply `requiredPermissions` to.

The clientside flow goes like this:
  1. Grain makes `postMessage` powerbox request on Alice's session, gets back a `requestToken` that has `introducerIdentity` set to Alice's identity ID.
  2. Grain calls `SandstormApi.claimRequest(requestToken, requiredPermissions)` on the server side. The returned capability has a `permissionsHeld` requirement so that it will stop working when Alice no longer holds `requiredPermissions`.
  3. The live capability gets saved. Note that here `grain`, `webkey` and `user` are all possibilities for the owner.

In a hypothetical server-side-only flow, the first two steps are merged into a single `SessionContext.request(query, requiredPermissions)` call.

It's the sturdyref created in step 3 that I think you're talking about showing up in an audit UI. The `requirements` field will hold the needed information. It's true that a capability can accumulate a large number of membrane requirements if it is passed through many powerbox requests. However, the way we represent that is through a chain of parent tokens; each link along the way will contain only the requirements picked up during the powerbox interaction which minted that token.

zarvox commented:
It is not obvious to me if or why I should assume that some (the last?) item in the `requirements` array is, in fact, the introducer's identity, should I wish to show that to the user.

If we do make that guarantee, we should probably document it and assert about it in the code.   I have always assumed `requiredPermissions`/`requirements` are an implementation detail of the permissions computation algorithm, rather than something with a guaranteed layout suitable for presentation to a user.  Happy to have that clarified, though!

This suggests that we will need some *pretty complicated* join logic to pull profiles for identities referenced by any number of ApiTokens, but perhaps that was already true.

zarvox commented:
One issue that I'm not sure this addresses is "how do I associate an identity (not just an account) with an `IpNetwork` I've granted to a grain, so I can show a sane thing in the UI?"  There, the `Supervisor.MembraneRequirement` is just `userIsAdmin: accountId`.

Maybe userIsAdmin needs to be a struct of (accountId, identityId), rather than just a Text?

dwrensha commented:
> There, the Supervisor.MembraneRequirement is just userIsAdmin: accountId.

To pass the `IpNetwork` to a grain, you need to go through a powerbox flow, which must be associated to a `SessionContext`, which will have a particular identity associated with it, and should attach a `permissionsHeld` requirement.

zarvox commented:
Ahhh, so for a hypothetical `IpNetwork` in the grain-specific audit interface, I'll look only for tokens with `owner.grain` with the matching grainId, and can always infer the introducer identity from the `requirements`.

Okay, so it sounds like the guarantee is:

* An ApiToken with owner `grain` is guaranteed to have as the last entry in its `requirements` a `permissionsHeld` struct which represents the account/identity responsible for creating the connection.

Does that sound right?

(Also, thanks for helping me understand all the intent here!)

dwrensha commented:
> An ApiToken with owner grain is guaranteed to have as the last entry in its requirements a permissionsHeld struct which represents the account/identity responsible for creating the connection.

Note quite. I would say that the entire `requirements` list is relevant, not just the final element in the list.

Consider the following interaction.
  1. Admin Alice grants an `IpNetwork` to grain G, which saves the cap as a sturdyref T with a `requirements` list including "Alice's account has admin privileges" and "Alice can access G".
  2. Alice shares G to Bob.
  3. Bob opens grain H, which makes a powerbox request for an `IpNetwork`. Bob chooses to allow G to fulfill the request, and to do that G uses the sturdyref T from step 1.
  4. H saves its new `IpNetwork` the resulting sturdyref has a `requirements` list consisting only of the `permissionsHeld` requirement "Bob can access H". The new sturdyref U points to T as its parent token.

The `IpNetwork` cap has passed through several powerbox interactions. At each step along the chain the local `requirements` list consists only of those requirements relevant to the most recent powerbox interaction. But to restore the cap, all of the requirements along the chain of parent tokens must be fulfilled.

The "requirements relevant to the most recent powerbox interaction" seems to me to be exactly the information needed for a an audit UI. I suppose it is plausible that in future revisions of the powerbox this may no longer be the case (e.g. if we support capabilities that don't autorevoke when the sharer loses access to the intermediary grain). Then we will need to add some additional way to capture the data needed in the audit UI. But I say let's not worry about that until it becomes a problem.

dwrensha commented:
After some updates to the copy/paste powerbox flow and [these changes](https://github.com/jparyani/sandstorm-test-app/pull/2) to the powerbox test app, the integration tests are now passing.

dwrensha commented:
Note that I'm not quite happy with `ViewInfo.getIconUrls()`. As discussed in https://github.com/sandstorm-io/sandstorm/pull/2016#discussion_r66898963, maybe this should be `ViewInfo.getIconAssetIds()` and we should add a `staticHost` field to `WebSession.Params`.

dwrensha commented:
In the latest commit, `ViewInfo` gets a `metadata: DenormalizedGrainMetadata` field and `WebSession.Params` gets a `staticAssetPath: Text` field. This allows a collections app to read icon asset IDs and then turn the into URLs suitable for including in an `<img>` tag. Each appearance of an asset ID in a `Grains.cachedViewInfo` increments that asset's reference count.

This approach still does not seem quite satisfactory, because a grain has no way either to prevent an asset from being deleted or even to get notified when an asset is deleted, short of repeatedly calling `getViewInfo()`. This suggests to me that we should try something like the following:

```capnp
interface StaticAsset {
    getUrl @0 () -> (url :Text);
}

 SandstormApi {
    // ....
    createStaticAsset @8 (mimeType :Text, encoding :Text, content :Data)
        -> (asset :StaticAsset);
}

struct GrainMetadata {
    appTitle @0 :LocalizedText;
    union {
         icon :group {
             format @1 :Text;
             assetId @2 :StaticAsset;
             assetId2xDpi @3 :StaticAsset;
             # If present, the asset ID for the equivalent asset as assetId at twice-resolution
         }
         appId @4 :Text;
    }
}

struct ViewInfo {
    /// ...
    metadata @5 :GrainMetadata;
} 

```

That is, we would put the actual `StaticAsset` capabilities in the `ViewInfo`. Then a grain would be able to prevent an asset from being deleted, simply by holding on to a reference to it; when a `StaticAsset` is restored, it increments the refcount in the database, and when it is dropped, it decrements the refcount.

I think we would also want to prevent grains from providing their own implementations of `StaticAsset`, to avoid possible cross-site scripting attacks. I think such prevention should be possible via a `capnp::MembranePolicy`.

@kentonv: I'm interested in your thoughts on all of this.

dwrensha commented:
Another idea for keeping assets valid: we could add a method like

```capnp
interface SandstormApi {
    // ....
    holdStaticAsset @9 (assetId :Text) -> (handle :Util.Handle);
}
```

where the asset's refcount is incremented for as long as `handle` is held. This would allow us to pass around asset IDs like in my proposed `ViewInfo.metadata`. A grain that wants to use an asset calls `holdStaticAsset()` to check that it's still valid and to ensure that it stays valid.

kentonv commented on shell/server/core.js:
```diff
@@ -176,17 +176,40 @@ class NotificationHandle {
+      }
+
+      if (!viewInfo.grainIconUrl) {
+        // TODO(security, now): Do we need to do some kind of filtering here to prevent
```
This TODO wouldn't be caught by our release script, FWIW. You need to do `TODO(now) TODO(security)`, or just say that `TODO(now)` takes precedence.

kentonv commented on shell/server/core.js:
```diff
@@ -158,17 +187,46 @@ class NotificationHandle {
-  return new Capnp.Capability(new PersistentUiViewImpl(persistentMethods), PersistentUiView);
+const makePersistentUiView = function (persistentMethods, grainId) {
+  check(grainId, String);
+  const grain = Grains.findOne({ _id: grainId });
```
Why not do this query on-demand when `getViewInfo()` is invoked? Seems like the view info can become outdated if someone restores the capability and keeps it around for a while.

dwrensha replied:
Good point. I think makes sense to also do the query here so we can throw an error early if the grain has been deleted.

kentonv commented on src/sandstorm/supervisor.capnp:
```diff
@@ -274,15 +256,28 @@ struct ApiTokenOwner {
-      # to `restore()` will be checked. This field is only intended to be filled in by the
-      # front-end during a powerbox request; a regular `save()` call produces a capability that
-      # has no "introducer".
+      # token. The token will become invalid if this identity loses access to the grain. This field
```
Is there a good reason not to declare `grain.introducerIdentity` entirely obsolete and ignore it?

zarvox replied:
I want an introducer identity to visualize which admin granted a grain an admin capability -- right now you can't get that from just a `userIsAdmin` membrane requirement.

I'd be okay with `userIsAdmin` turning from a `Text` into a struct or group with an `accountId` and an `identityId` and saying the requirements are the canonical location for this information.  But the issuing identity needs to be stored somewhere.

kentonv replied:
I believe that an admin capability will, in the new model, always have two `MembraneRequirements` -- a `userIsAdmin` and a `permissionsHeld`. The latter tells you the identity under which the user was acting when they injected the capability into the grain.

zarvox replied:
Okay, that works for me!

dwrensha replied:
> Is there a good reason not to declare `grain.introducerIdentity` entirely obsolete and ignore it?

It would be nice if we could continue to support existing grain-owned tokens. Most of these will have an introducer identity and therefore should be auto-revoked if that identity loses access to the grain. To accomplish this, we could either migrate these tokens so that the data in `introducerIdentity` field moves into the `requirements` field, or we could follow my strategy here, which is to continue honoring the `grain.introducerIdentity` field if it is present.

zarvox replied:
FWIW, I personally tend to favor migrations where possible over needing to remember to support both structures forever.  It turns an ongoing maintenance cost/conceptual debt (how should the code behave when it encounter tokens that have this shape + educating/reminding devs that once upon a time we didn't enforce the invariants that new tokens uphold) into one that can be paid down all-at-once.

I think it's higher cost to teach future devs (including my future self, who will likely have forgotten) about various inconsistencies than it is to remove them from existence.

That said, I'm also wary of asking other people to write possibly tricky code, and it seems the other implementation already lives, so I'm reasonably happy with either of David's suggestions, and if I want to put my money where my mouth is, I could write a migration for this at ~any time in the future.

kentonv replied:
@dwrensha The problem is existing tokens were not only expecting `introducerIdentity` to have basic access but were also expecting to check it against some required permissions specified at restore() time. The new interface doesn't give any way to specify said required permissions. So, you are loosening the requirements here either way. Why is it OK to loosen them somewhat but not completely?

I think the only self-consistent options here are:
1. Support exactly what we had before, including keeping a way to specify requiredPermissions on restore.
2. Declare that because no one was actually using these yet (except feature-key-vendor), we don't care. In this case probably we should refuse to restore any token that has `introducerIdentity` at all, on the basis that we can no longer correctly enforce whatever it was trying to express. Or write a migration that deletes these tokens.

I vote for 2.

dwrensha replied:
> Why is it OK to loosen them somewhat but not completely?

I was somehow under the impression that any apps that do use the powerbox today would not yet need full support for `grain.requiredPermissions`. But apparently that's not the case even for the feature key vendor app.

I agree that we should refuse to restore such tokens.

kentonv commented on shell/packages/sandstorm-db/db.js:
```diff
@@ -1810,23 +1810,81 @@ if (Meteor.isServer) {
-        Url.parse(process.env.ROOT_URL).protocol + "//" + this.makeWildcardHost("static") + "/";
+  SandstormDb.prototype.updateCachedViewInfo = function (grainId, viewInfo) {
+    check(grainId, String);
+    // TODO(soon): Typecheck viewInfo. We should be able to factor out some of the checking code
```
I guess type checking viewInfo isn't critical because we normally receive it via Cap'n Proto, which already enforces the type, right? Mention that here.

kentonv commented on shell/packages/sandstorm-db/db.js:
```diff
@@ -1810,23 +1810,81 @@ if (Meteor.isServer) {
-                              assetId2xDpi: Match.Optional(String), }));
-    const root =
-        Url.parse(process.env.ROOT_URL).protocol + "//" + this.makeWildcardHost("static") + "/";
+  SandstormDb.prototype.updateCachedViewInfo = function (grainId, viewInfo) {
```
Probably worth adding some doc comment explaining what this method is doing.

I guess all the refcounting here only happens if the app defines grain-specific overrides for the icons, which it usually doesn't, right?

dwrensha replied:
The refcounting happens when the grain specifically fills in `ViewInfo.metadata`. I wouldn't want to rule out the possibility that this could become the usual case in the future, e.g. if it becomes common for apps to provide multiple types of actions and to present a different grain icon for each action.

kentonv commented:
OK, reviewed everything.

I'm worried that referencing assets by string ID rather than by capability is going to create obstacles somewhere down the road, particularly when we start having federation between Sandstorm servers where the same asset probably has different IDs on different servers.

I think we will want to be aware when asset references move between servers, so that data can automatically be synced in the background, etc. We obviously only have that ability with capabilities.

Let me think about this more tomorrow.

kentonv commented:
As discussed in our meeting, let's take the simple approach for now of wrapping the asset IDs in capabilities that *aren't* persistable. The only thing you can do with these capabilities is call a further method to obtain the asset ID. The purpose here is to give us the opportunity to extend the behavior later without breaking backwards compatibility.

For now, we probably should not allow an app to specify its own assets, since I'm not sure currently how that would work. We can add that ability later when it is needed.

dwrensha commented:
Ready for re-review.

In this version, the `StaticAsset` capability can represent either some data in the `StaticAssets` collection or an identicon that will be served from an "identicon" host, which acts similarly to the "static" host, but does not require any backing data. This idea came up in a separate discussion: https://github.com/sandstorm-io/sandstorm/issues/1866#issuecomment-228872122.

kentonv commented on src/sandstorm/grain.capnp:
```diff
@@ -160,27 +160,11 @@ interface SandstormApi(AppObjectId) {
-    appId @4 :Text;
-    # App ID, needed to generate a favicon if no icon is provided.
-  }
+  getUrl @0 () -> (url :Text);
```
This is XSS-prone: If the StaticAsset capability is implemented by another app or other malicious source, it can return a `javascript:` URL or similar, which is likely to be evaluated in the context of the calling app. You can say it is the caller's responsibility to filter, but most won't.

We can prevent this problem by changing the return to something like:

    (https: Bool, hostPath :Text)

The URL is then formed as:

    (https ? "https://" : "http://") + hostPath

kentonv commented on shell/server/pre-meteor.js:
```diff
@@ -452,9 +460,9 @@ Meteor.startup(() => {
       }
 
-      if (id === 'static') {
+      if (id === "static" || id === "identicon") {
```
There's an argument to be made that we should differentiate these by path rather than by host: When a browser sees a new hostname, it is generally forced to create a new HTTP/HTTPS connection to that host rather than reuse an existing one, hurting performance.

dwrensha commented:
I've split the return value of `StaticAsset.getUrl()` into a `protocol` and a `hostPath` field, and I've moved the identicon server to our existing "static" host, under the "/identicon/" path.

Ready for re-review.
